grammar/rainerscript.c

```diff
@@ -2122,6 +2122,50 @@ doFuncCall(struct cnffunc *__restrict__ const func, struct svar *__restrict__ co
 		if(bMustFree) free(str);
 		varFreeMembers(&r[1]);
 		break;
+	case CNFFUNC_FORMAT_TIME: {
+		time_t unixtime;
+		struct tm lt;
+		char   result[64];
+		char  *formatstr = NULL, *endptr = NULL;
+
+		cnfexprEval(func->expr[0], &r[0], usrptr);
+		cnfexprEval(func->expr[1], &r[1], usrptr);
+
+		str = (char*) var2CString(&r[0], &bMustFree);
+		formatstr = (char*) es_str2cstr(r[1].d.estr, NULL);
+
+
+		// TODO: Error checking, and account for different
+		//       possible sizes of time_t 32/64 bit.
+		errno = 0;
+		unixtime = (time_t) strtoll(str, &endptr, 10);
```

Comment:
There is a specific function to obtain a number from the script processor. This should be used.

Reply:
Riiiggt, I'm still a bit of n00b on the RainerScript internals. Let me review. My intent is to allow fractional seconds to be passed in where I split off of the decimal (if present). I see `var2Number()` and `str2num()`, and they look like they'd parse up to the decimal and stop, but they'd not tell me where the decimal is, where strto?l? would...

Reply:
Ah, I see. We do **not** have fractional numbers in a unix timestamp. The ElasticSearch ms timestamp is as follows: let t be the regular unix timestamp and f the fractional part. Then you will see timestamp = t*1000+f, so it always is an integer.

I would strongly suggest to stick with that definition. As an alternative, the right thing would be to use the syslog-internal timestamp format, which has all fields nicely split of (including fractional part, precisio, TZ). But shuffling this is **a lot** of work and it is not compatible with anything (including incompatibility with current liblognorm). I think the ES approach is indeed a good way to do it.

Reply:
True, but I have seen unix-like timestamps with thousandths after a decimal which is what I was trying to accommodate. Maybe I'm getting carried away :)

I have no problem saying UNIX timestamps are non-fractional, and that's that. It's easier to support. I'll forget about it for now - if someone in the future asks for that sort of thing, we can review it then.

Reply:
Let's say it's non fractional. There are a myriad of different formats. Once this all works, I would suggest to review this again. I think we can also probably apply the same heuristics ES uses (at least I assume so). For any recent timestamp, you can detect if it is with or without fractional seconds just by looking at the value (as the ms timestamp is 1000 times as high). Doesn't work for timestamp back in the 1970s, but that should no longer be a concern in our context.

But for now I'd just use standard format and get that solidly done.